### PR TITLE
JAMES-3113 Mails send from an alias should be considered as localSender

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by Arsnael** [comment]
```
[e43e66b682c2128b143cc2764ad5af8006aaaf3e] [ERROR] Errors: 
[e43e66b682c2128b143cc2764ad5af8006aaaf3e] [ERROR]   JamesMailetContextTest.isLocalEmailShouldBeFalseWhenUsernameDoNotExistButDomainExists:160 » IllegalState
[e43e66b682c2128b143cc2764ad5af8006aaaf3e] [ERROR]   JamesMailetContextTest.isLocalUserShouldReturnFalseWhenUsedWithLocalPartAndUserDoNotExistOnDefaultDomain:143 » IllegalState
```

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SenderIsLocalIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import org.apache.james.webadmin.routes.AliasRoutes;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import io.restassured.specification.RequestSpecification;
+
+public class SenderIsLocalIntegrationTest {
```
Comment (Arsnael) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SenderIsLocalIntegrationTest.java]: new test class, JUnit 5?

**Comment by chibenwa** [comment]
test this please

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SenderIsLocalIntegrationTest.java (hunk 1 above)
Comment (chibenwa) [reply server/mailet/integration-testing/src/test/java/org/apache/james/mailets/SenderIsLocalIntegrationTest.java]: No, that's a huge effort in this project

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailetContext.java [hunk 2]
```diff
@@ -248,14 +257,27 @@ public boolean isLocalEmail(MailAddress mailAddress) {
...
             }
         }
         return false;
     }
 
+    private boolean isLocaluser(MailAddress mailAddress) throws UsersRepositoryException {
+        return localusers.contains(localusers.getUser(mailAddress));
+    }
+
+    private boolean isLocalAlias(MailAddress mailAddress) throws UsersRepositoryException, ErrorMappingException, RecipientRewriteTableException {
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailetContext.java]: This method would probably deserve its own class, JamesMailetContext is already bloated and does too many things, let's not make it bigger.

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+
+public class LocalResources {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LocalResources.class);
+    private static final EnumSet<Mapping.Type> ALIAS_TYPES = EnumSet.of(Mapping.Type.Alias, Mapping.Type.DomainAlias);
+
+    private final UsersRepository localusers;
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java]: localUsers

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    private final DomainList domains;
+    private final RecipientRewriteTable recipientRewriteTable;
+
+    @Inject
+    public LocalResources(UsersRepository localusers, DomainList domains, RecipientRewriteTable recipientRewriteTable) {
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java]: why do you need it to be public?

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        this.domains = domains;
+        this.recipientRewriteTable = recipientRewriteTable;
+    }
+
+    public boolean isLocalServer(Domain domain) {
+        try {
+            return domains.containsDomain(domain);
+        } catch (DomainListException e) {
+            LOGGER.error("Unable to retrieve domains", e);
+            return false;
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java]: this is a bad fallback, not being able to do a lookup should not end in considering a domain to not be handle by the server.

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        }
+    }
+
+    public boolean isLocalUser(String name) {
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java]: what about creating a Username from this string?

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        if (name == null) {
+            return false;
+        }
+        try {
+            if (!name.contains("@")) {
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java]: we should not (only) decide based on whether it contains an `@` but also take care of virtual hosting, don't you think?

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                try {
+                    return isLocalEmail(new MailAddress(name.toLowerCase(Locale.US), domains.getDefaultDomain().asString()));
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java]: this code is duplicate of what Username do

File: server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                } catch (DomainListException e) {
+                    LOGGER.error("Unable to access DomainList", e);
+                    return false;
```
Comment (mbaechler) [inline server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/LocalResources.java]: this fallback looks wrong: we will eventually take a bad decision based on a technical error

**Comment by chibenwa** [comment]
> I did a quick review of the extracted code, I think there are real bugs in it but feel free to address these problems in another PR.

I'm pretty sure of this.

Some of this code is never called.

IMO to be addresses in a follow up PR

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/3210 for further fixing the JamesMailetContext

**Comment by Arsnael** [comment]
```
[c9116123ed987604fe8e6d5fc1b30215ad53caa4] [ERROR]   JamesSpringContextTest.springShouldLoadAndAddOnlyOneQuotaUpdaterListener Multiple Failures (2 failures)
[c9116123ed987604fe8e6d5fc1b30215ad53caa4] 	org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'mailprocessor' defined in class path resource [META-INF/spring/mailetcontainer-context.xml]: Unsatisfied dependency expressed through constructor argument with index 1 of type [org.apache.mailet.MailetContext]: : Error creating bean with name 'mailetcontext' defined in class path resource [META-INF/spring/mailetcontainer-context.xml]: Unsatisfied dependency expressed through constructor argument with index 2 of type [org.apache.james.mailetcontainer.LocalResources]: : No qualifying bean of type [org.apache.james.mailetcontainer.LocalResources] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.apache.james.mailetcontainer.LocalResources] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'mailetcontext' defined in class path resource [META-INF/spring/mailetcontainer-context.xml]: Unsatisfied dependency expressed through constructor argument with index 2 of type [org.apache.james.mailetcontainer.LocalResources]: : No qualifying bean of type [org.apache.james.mailetcontainer.LocalResources] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.apache.james.mailetcontainer.LocalResources] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}
[c9116123ed987604fe8e6d5fc1b30215ad53caa4] 	java.lang.NullPointerException: <no message>
[c9116123ed987604fe8e6d5fc1b30215ad53caa4] [ERROR] Errors: 
[c9116123ed987604fe8e6d5fc1b30215ad53caa4] [ERROR]   JamesAppSpringMainTest.testServer:27 » UnsatisfiedDependency Error creating be...
```

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**